### PR TITLE
commit of verification long standing branch

### PR DESCRIPTION
**Pull request by andreistefanescu**

### CODE REVIEW

**Comment by dwightguth** [comment]
I don't see anything obviously wrong with this but I'm tired and not focused and don't know the rewriter very well. @yilongli can you give me a sanity check here, see if you see anything substantial that feels weird to you?

**Comment by yilongli** [comment]
I will do a thorough review immediately.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java [hunk 1]
```diff
@@ -20,38 +22,71 @@
...
+        return mapUpdate.evaluateUpdate();
+    }
 
-        BuiltinMap.Builder builder = BuiltinMap.builder();
-        builder.putAll(term.getEntries());
-        builder.put(key, value);
+    public static Term updateAll(BuiltinMap map1, BuiltinMap map2, TermContext context) {
+        BuiltinMap.Builder builder = new BuiltinMap.Builder();
+        builder.concatenate(map1, map2);
         return builder.build();
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:63]: Why are there `mapUpdate.evaluateUpdate()`'s at the end of the above methods but not this one?

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:63]: I see, never mind... I guess I need to do the code review in an IDE.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java [hunk 2]
```diff
@@ -20,38 +22,71 @@
...
+        return (BuiltinSet) builder.build();
+    }
+
+    public static BuiltinList values(BuiltinMap map, TermContext context) {
+        if (!map.isConcreteCollection()) {
+            return null;
+        }
+
+        List<Term> elements = new ArrayList<>(map.getEntries().values());
+        return new BuiltinList(elements);
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:82]: `elements` will be copied again in the constructor of `BuiltinList`

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java [hunk 3]
```diff
@@ -1,74 +1,76 @@
...
 import org.kframework.backend.java.symbolic.Visitor;
 import org.kframework.backend.java.util.Utils;
 import org.kframework.kil.ASTNode;
-import org.kframework.kil.DataStructureSort;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.collections15.map.UnmodifiableMap;
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java]: `org.apache.commons.collections4.map.UnmodifiableMap` is actually the newer version. The only reason I couldn't remove `org.apache.commons.collections15` is that the API's of JUNG library depend on it explicitly.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+/**
+ * Table of {@code public static} methods common to multiple builtin collections.
+ *
+ * @author AndreiS
+ */
+public final class BuiltinCollectionOperations {
+
+    private BuiltinCollectionOperations() { }
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java:16]: dead code?

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java (hunk 2 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:82]: I know... I did not not change `BuiltinList` in this commit.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinSetOperations.java [hunk 5]
```diff
@@ -17,38 +15,56 @@
...
-    public static BuiltinSet construct(BuiltinSet term1, BuiltinSet term2, TermContext context) {
-        Variable frame = null;
-        if (term1.hasFrame() && term2.hasFrame()) {
-            throw new IllegalArgumentException(
-                    "both set arguments have frames, but the combined set cannot have two frames");
-        } else if (term1.hasFrame()) {
-            frame = term1.frame();
-        } else if (term2.hasFrame()) {
-            frame = term2.frame();
+    public static Term construct(BuiltinSet set1, BuiltinSet set2, TermContext context) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinSetOperations.java]: in the `hook.properties`, you actually renamed `construct` to `constructor`.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java (hunk 2 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:82]: so you will also separate `ListUpdate` from `BuiltinList` in future commit?

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinSetOperations.java (hunk 5 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinSetOperations.java]: correct

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java [hunk 6]
```diff
@@ -114,23 +120,19 @@ protected boolean computeHasCell() {
 
     @Override
     public String toString() {
-        return toString(" ", " |-> ", DataStructureSort.DEFAULT_MAP_UNIT_LABEL);
+        return toString(" ", " |-> ", ".Map");
     }
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java:118]: extract a constant out of the string literal?

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java [hunk 7]
```diff
@@ -20,38 +22,71 @@
  */
 public class BuiltinMapOperations {
 
-    public static BuiltinSet keys(BuiltinMap term, TermContext context) {
-        Preconditions.checkArgument(!term.hasFrame(), "argument " + term + " has frame");
+    public static Term construct(BuiltinMap map1, BuiltinMap map2, TermContext context) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java]: construct -> constructor

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java]: can you go over the hook.properties file to make sure there are no other bugs of the this kind?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java [hunk 8]
```diff
@@ -193,62 +183,56 @@ public Term remove(Term key) {
         }
 
         public Map<Term, Term> getEntries() {
-            return UnmodifiableMap.unmodifiableMap(entries);
+            return UnmodifiableMap.decorate(entries);
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java]: Once you switch back to the Apache commons collection 4.0, it would be `unmodifiableMap`

File: src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+     * TODO: improve name/description
+     */
+    public boolean isConcreteCollection() {
+        return collectionPatterns.isEmpty()
+                && collectionVariables.isEmpty()
+                && collectionFunctions.isEmpty();
+    }
+
+    @Override
+    public boolean isExactSort() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java]: not a big deal, but I would like to make it `final`

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java [hunk 10]
```diff
@@ -109,7 +133,7 @@ public static boolean matchable(Term subject, Term pattern, TermContext context)
      *         variables in the pattern to sub-terms in the subject)
      */
     public static List<Map<Variable, Term>> patternMatch(Term subject, Rule rule, TermContext context) {
-        PatternMatcher matcher = new PatternMatcher(context);
+        PatternMatcher matcher = new PatternMatcher(rule.containsAttribute("lemma"), context);
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java:140]: use `Attribute.LEMMA_KEY`?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java [hunk 11]
```diff
@@ -199,7 +223,7 @@ public static boolean matchable(Term subject, Term pattern, TermContext context)
                         // the non-lookup term is not a variable and thus requires further pattern matching
                         // for example: L:List[Int(#"0")] = '#ostream(_)(I:Int), where L is the output buffer
                         //           => '#ostream(_)(Int(#"1")) =? '#ostream(_)(I:Int)
-                        PatternMatcher lookupMatcher = new PatternMatcher(context);
+                        PatternMatcher lookupMatcher = new PatternMatcher(rule.containsAttribute("lemma"), context);
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java]: ``` Attribute.LEMMA_KEY```
```

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java [hunk 12]
```diff
@@ -385,7 +390,9 @@ private ConstrainedTerm constructNewSubjectTerm(Rule rule, SymbolicConstraint co
                 constraint.substitution(),
                 constraint.termContext());
         /* eliminate anonymous variables */
-        constraint.eliminateAnonymousVariables();
+        constraint.eliminateAnonymousVariables(existingVariables);
+        // TODO(AndreiS): functions not being evaluated is becoming quite annoying
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java:394]: I agree. I had to comment out an assertion in `LocalEvaluator#transform(KItem)` because it failed the `visitor/imp.k` test. Do you think that assertion should be enforced?

**Comment by yilongli** [comment]
I have roughly go through all the changes. It looks good to me except for a few things that I have commented on. I will need to read them more carefully later when I start to prove some small examples.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java (hunk 12 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java:394]: I think we should have a isNormal flag for each term, and that a term should always be in normal form. It should not be that difficult to maintain and use when resolving functions (and now anywhere rules), but it is not part of this commit. We should discuss it separately.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java (hunk 2 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinMapOperations.java:82]: Yes.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java (hunk 4 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java:16]: If you are referring to the private empty constructor, this is common for classes with only static fields/methods, but I don't remember the reason. @dwightguth what was it?

**Comment by andreistefanescu** [comment]
Yeah...I merged a change in cil which only works with Radu's branch. @dwightguth what is the policy here?

File: include/java/hooks.properties [hunk 13]
```diff
@@ -1,215 +1,231 @@
...
+\#STRING\:findChar : org.kframework.backend.java.builtins.BuiltinStringOperations.findChar
+\#STRING\:rfindChar : org.kframework.backend.java.builtins.BuiltinStringOperations.rfindChar
+\#STRING\:categoryChar : org.kframework.backend.java.builtins.BuiltinStringOperations.category
+\#STRING\:directionalityChar : org.kframework.backend.java.builtins.BuiltinStringOperations.directionality
+\#STRING\:replaceAll : org.kframework.backend.java.builtins.BuiltinStringOperations.replaceAll
+\#STRING\:replace : org.kframework.backend.java.builtins.BuiltinStringOperations.replace
+\#STRING\:replaceFirst : org.kframework.backend.java.builtins.BuiltinStringOperations.replaceFirst
+\#STRING\:countAllOccurences : org.kframework.backend.java.builtins.BuiltinStringOperations.countAllOccurences
+
+\#MINT\:constructor : org.kframework.backend.java.builtins.BuiltinBitVectorOperations.construct
```
Comment (yilongli) [inline include/java/hooks.properties:103]: you might want to rename the method to `constructor`?

File: samples/java_rewrite_engine/tutorial/2_languages/1_simple/1_untyped/verification/reverse_spec.k [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        (list(X1)(B:Seq) list(Y1)(C:Seq) => list(?Y2)(?A:Seq))
+    ...</store>
+    <busy>
+        _:Set
+    </busy>
+    <terminated>
+        _:Set
+    </terminated>
+    <in>
+        // TODO(AndreiS): inverstigate why I/O is not disabled in prove
```
Comment (yilongli) [inline samples/java_rewrite_engine/tutorial/2_languages/1_simple/1_untyped/verification/reverse_spec.k:106]: you may want to take a look at `KRunOptions#io()` and `KRunOption.Experimental#prove()`

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java (hunk 4 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinCollectionOperations.java:16]: I see. You want to forbid instantiating this class.

**Comment by yilongli** [comment]
When I read your code related to builtin operations, I realize something I did in the past was problematic. Although it's not part of this commit, can you help me fix `MetaK#unifiable` or put a TODO comment with my name? I think it should return `null` when the constraint cannot be solved.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+    public ImmutableMultiset<Variable> collectionVariables() {
+        return collectionVariables;
+    }
+
+    /**
+     * Returns true if this collection consists only of elements or entries, but no patterns, functions or variables.
+     * TODO: improve name/description
+     */
+    public boolean isConcreteCollection() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java]: final?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+                && collectionFunctions.isEmpty();
+    }
+
+    @Override
+    public final boolean isExactSort() {
+        return true;
+    }
+
+    @Override
+    public boolean isGround() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java]: final

**Comment by yilongli** [comment]
Regarding `final` modifiers, you don't have to change them if you don't want to. I like to have it because it makes the code look more clear to me, especially when there are several subclasses. I can change them later when I start to study the code.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        return isConcreteCollection() && super.isGround();
+    }
+
+    @Override
+    public boolean isLHSView() {
+        // TODO(AndreiS): this method should be removed in later commits
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java:74]: why? because it will be replaced by `isUnifiableByCurrentAlgorithm`?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java [hunk 18]
```diff
@@ -320,4 +320,12 @@ public int removeRight() {
         return removeRight;
     }
 
+    @Override
+    public final boolean isSymbolic() {
+        return hasFrame()
+                && elementsLeft.isEmpty()
+                && elementsRight.isEmpty()
+                && removeLeft == 0
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:328]: it looks a little weird to me, but I guess it will be completely changed once you separate the ListUpdate out of it.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java [hunk 19]
```diff
@@ -15,60 +9,62 @@
...
-    public Map<Term, Term> getEntries() {
-        return UnmodifiableMap.unmodifiableMap(entries);
+    public UnmodifiableMap<Term, Term> getEntries() {
+        return entries;
     }
 
-    @Override
-    public Iterator<Map.Entry<Term, Term>> iterator() {
-        return entries.entrySet().iterator();
+    public boolean isUnifiableByCurrentAlgorithm() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinMap.java:56]: final?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java [hunk 20]
```diff
@@ -81,7 +81,7 @@ public boolean isEmpty() {
     public abstract boolean isLHSView();
 
     @Override
-    public final boolean isSymbolic() {
+    public boolean isSymbolic() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Collection.java:84]: if you remove the `final` modifier, I think it's better to make it `abstract` to force all the subclasses properly implement this method.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java [hunk 21]
```diff
@@ -312,8 +312,12 @@ public boolean isFunction() {
         return super.containsAttribute(Attribute.FUNCTION_KEY);
     }
 
-    public KLabelConstant functionKLabel() {
-        assert isFunction();
+    public boolean isPattern() {
+        return super.containsAttribute(Attribute.PATTERN_KEY);
+    }
+
+    public KLabelConstant definedKLabel() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/Rule.java:322]: javadoc? I don't really have a good suggestion, but `definedKLabel` doesn't seem very straightforward to me. So I think it is good to at least have a javadoc.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java [hunk 22]
```diff
@@ -72,6 +79,7 @@ private KLabelConstant(String label, Definition definition) {
                 Production fstProd = iterator.next();
                 isFunction = fstProd.containsAttribute(Attribute.FUNCTION.getKey())
                         || fstProd.containsAttribute(Attribute.PREDICATE.getKey());
+                isPattern = fstProd.containsAttribute("pattern");
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java]: Attribute.PATTERN_KEY?

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java [hunk 23]
```diff
@@ -87,6 +95,10 @@ private KLabelConstant(String label, Definition definition) {
                         + label
                         + " is a function symbol because there are multiple productions associated with this KLabel: "
                         + productions;
+                assert isPattern == production.containsAttribute("pattern") :
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java]: Attribute.PATTERN_KEY

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelInjection.java [hunk 24]
```diff
@@ -68,6 +68,11 @@ public final boolean isFunction() {
     }
 
     @Override
+    public boolean isPattern() {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelInjection.java]: final?

**Comment by yilongli** [comment]
I think you forgot to change `BackendJavaKILtoKILTransformer#transform(BuiltinMap/BuiltinSet)` to take into consideration the newly added fields.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/BottomUpVisitor.java [hunk 25]
```diff
@@ -38,10 +38,19 @@ public void visit(BuiltinList node) {
 
     @Override
     public void visit(BuiltinMap builtinMap) {
-        for (java.util.Map.Entry<Term, Term> entry : builtinMap) {
+        for (java.util.Map.Entry<Term, Term> entry : builtinMap.getEntries().entrySet()) {
             entry.getKey().accept(this);
             entry.getValue().accept(this);
         }
+        for (KItem pattern : builtinMap.collectionPatterns()) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/BottomUpVisitor.java]: what about make a new method in `AssociativeCommutativeCollection` which returns the union of all three collections? In this way, you have shorter code in `Visitor`s and `Transformer`s. Besides, next time you want to change the contents of `AssociativeCommutativeCollection`, you don't have to go through many different places to make the change.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/CopyOnWriteTransformer.java [hunk 26]
```diff
@@ -347,67 +347,58 @@ public ASTNode transform(BuiltinMap builtinMap) {
...
-        } else {
-            return builtinSet;
+        boolean changed = false;
+        BuiltinSet.Builder builder = BuiltinSet.builder();
+        for(Term element : builtinSet.elements()) {
+            Term transformedElement = (Term) element.accept(this);
+            builder.add(transformedElement);
+            changed = changed || (transformedElement != element);
+        }
+        for (KItem kItem : builtinSet.collectionPatterns()) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/CopyOnWriteTransformer.java]: return all three collections in one method?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/BottomUpVisitor.java (hunk 25 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/BottomUpVisitor.java]: e.g. `AssociativeCommutativeCollection#baseTerms` to be consistent with the generic KIL?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java [hunk 27]
```diff
@@ -505,15 +448,22 @@ public ASTNode visit(org.kframework.kil.MapUpdate node, Void _)  {
 
     @Override
     public ASTNode visit(org.kframework.kil.Variable node, Void _)  {
+        String name = node.getName();
+        if (node.isFreshVariable()) {
+           name = "?" + name;
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java]: make a method in `Variable` to return this for you?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java [hunk 28]
```diff
@@ -67,6 +90,8 @@
      */
     private boolean isStarNested;
 
+    private final boolean isLemma;
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java:93]: javadoc? e.g. specifies whether the pattern is the left-hand side of a lemma rule? and maybe rename it to `isLemmaPattern`?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java [hunk 29]
```diff
@@ -285,10 +288,30 @@ assert checkKindMisMatch(leftHandSide, rightHandSide):
...
+                leftHandSide = kList.get(0);
+                rightHandSide = kList.get(1);
+
+            }
+
             this.leftHandSide = leftHandSide;
             this.rightHandSide = rightHandSide;
         }
 
+        public boolean isTermEquality(Term leftHandSide) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java]: leftHandSide => term?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java [hunk 30]
```diff
@@ -285,10 +288,30 @@ assert checkKindMisMatch(leftHandSide, rightHandSide):
...
+            return leftHandSide instanceof KItem
+                    && ((KItem) leftHandSide).kLabel() instanceof KLabelConstant
+                    && ((KLabelConstant) ((KItem) leftHandSide).kLabel()).label().equals("'_==K_");
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java]: extract `'_==K_` to a constant? or make a constant KLabelConstant?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java [hunk 31]
```diff
@@ -699,13 +816,14 @@ public boolean implies(SymbolicConstraint constraint) {
             SymbolicConstraint right = implication.right;
             if (left.isFalse()) continue;
 
-            if (DEBUG) {
+            if (context.definition().context().globalOptions.verbose) {
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicConstraint.java]: in case you need it, there is also a debug option

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java (hunk 12 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java:394]: do you want me to implement this? or are we waiting for Cosmin's refactoring? I think you mentioned that at some point?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicUnifier.java [hunk 32]
```diff
@@ -229,8 +357,9 @@ public void unify(BuiltinMap builtinMap, Term term) {
 //        if (builtinMap.equals(BuiltinMap.EMPTY) && term.equals(BuiltinMap.EMPTY))
 //            return;
 
-        throw new UnsupportedOperationException(
-                "map matching is only supported when one of the maps is a variable.");
+        //throw new UnsupportedOperationException(
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicUnifier.java:360]: Is it better to throw an Exception when the pattern is not supported?

**Comment by yilongli** [comment]
OK, I just go through the changes again more carefully. I have nothing else to complain:)

**Comment by dwightguth** [comment]
@andreistefanescu the policy is that I will merge this even though CIL fails because we know why it is doing so and will be addressed in the near future.

**Comment by dwightguth** [comment]
Jenkins: test this please

**Comment by dwightguth** [comment]
Since I don't know how quickly Radu will be able to complete the remaining changes I asked for in his code review, I have fixed Jenkins to a specific old version of the c semantics repository so that we can continue to test CIL until he is ready to merge.

**Comment by andreistefanescu** [comment]
@dwightguth thank you for taking care of CIL

File: src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java (hunk 17 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/AssociativeCommutativeCollection.java:74]: Yes. Also, it used to be the case the collections in the lhs of rules would respect the isLHSView predicate. Now, this is no longer the case.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java (hunk 18 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/BuiltinList.java:328]: This is kind of hackish. But we need to change ListBuiltin anyway, so this code will change.

**Comment by andreistefanescu** [comment]
Fixed it.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java (hunk 12 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SymbolicRewriter.java:394]: We should discuss after you are done with your class. There are several
things that may need be implemented.

**Comment by andreistefanescu** [comment]
@yilongli thanks a lot for such detailed comments! I think I addressed all of them.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java [hunk 33]
```diff
@@ -48,7 +48,7 @@ public static BoolToken unifiable(Term term1, Term term2, TermContext context) {
         } catch (UnificationFailure e) {
             return BoolToken.FALSE;
         }
-        return constraint.isFalse() ? BoolToken.FALSE : BoolToken.TRUE;
+        return constraint.isSubstitution() ? BoolToken.TRUE : null;
```
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: if `constraint.isSubstitution` then I understand this method should return `BoolToken.TRUE`. But what if `constraint.isFalse` is true, shouldn't we return `BoolToken.FALSE` instead of null?

**Comment by yilongli** [comment]
I don't have other comments except for the one above. Thanks for addressing all these small issues!

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java (hunk 33 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: Hmm...true. Although, if you want to be thorough, you should also call simplify().
